### PR TITLE
Add public API for regex charset parser

### PR DESCRIPTION
**Pull request by DmitryOlshansky**
It's a bunch of unwieldy code that I planed to reuse long ago.
Libraries such as ICU also do provide means to construct sets of codepoints from strings with regex syntax.

### CODE REVIEW

**Comment by wilzbach** [comment]
@DmitryOlshansky 
FYI failing on the auto-tester

```d
generated/d_do_test.o(.text._D3std4conv29__T8textImplTAyaTAyaTkTAyaTkZ8textImplFNaNbNfAyakAyakZAya+0x3d): In function `_D3std4conv29__T8textImplTAyaTAyaTkTAyaTkZ8textImplFNaNbNfAyakAyakZAya':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o(.text._D3std4conv29__T8textImplTAyaTAyaTkTAyaTkZ8textImplFNaNbNfAyakAyakZAya+0x79): In function `_D3std4conv29__T8textImplTAyaTAyaTkTAyaTkZ8textImplFNaNbNfAyakAyakZAya':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o(.text._D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv+0x38): In function `_D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result5emptyMFNaNbNdNiNfZb'
generated/d_do_test.o(.text._D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv+0x4a): In function `_D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result5frontMFNaNbNdNiNfZa'
generated/d_do_test.o(.text._D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv+0x5b): In function `_D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result8popFrontMFNaNbNiNfZv'
generated/d_do_test.o(.text._D3std4conv17__T6toImplTAyaTkZ6toImplFNaNbNekkE3std5ascii10LetterCaseZAya+0x51): In function `_D3std4conv17__T6toImplTAyaTkZ6toImplFNaNbNekkE3std5ascii10LetterCaseZAya':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o(.text._D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa+0x10): In function `_D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result6lengthMFNaNbNdNiNfZk'
generated/d_do_test.o(.text._D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa+0x4b): In function `_D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result5emptyMFNaNbNdNiNfZb'
generated/d_do_test.o(.text._D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa+0x57): In function `_D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result5frontMFNaNbNdNiNfZa'
generated/d_do_test.o(.text._D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa+0x8c): In function `_D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result8popFrontMFNaNbNiNfZv'
generated/d_do_test.o(.text._D3std4conv36__T8textImplTAyaTAyaTkTAyaTykTAyaTkZ8textImplFNaNbNfAyakAyaykAyakZAya+0x3d): In function `_D3std4conv36__T8textImplTAyaTAyaTkTAyaTykTAyaTkZ8textImplFNaNbNfAyakAyaykAyakZAya':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o(.text._D3std4conv36__T8textImplTAyaTAyaTkTAyaTykTAyaTkZ8textImplFNaNbNfAyakAyaykAyakZAya+0xb5): In function `_D3std4conv36__T8textImplTAyaTAyaTkTAyaTykTAyaTkZ8textImplFNaNbNfAyakAyaykAyakZAya':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o(.text._D3std4conv30__T8textImplTAyaTAyaTkTAyaTykZ8textImplFNaNbNfAyakAyaykZAya+0x3d): In function `_D3std4conv30__T8textImplTAyaTAyaTkTAyaTykZ8textImplFNaNbNfAyakAyaykZAya':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o(.text._D3std4conv23__T8textImplTAyaTAyaTkZ8textImplFNaNbNfAyakZAya+0x3d): In function `_D3std4conv23__T8textImplTAyaTAyaTkZ8textImplFNaNbNfAyakZAya':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o(.text._D3std4conv27__T8textImplTAyaTAyaTkTAyaZ8textImplFNaNbNfAyakAyaZAya+0x3a): In function `_D3std4conv27__T8textImplTAyaTAyaTkTAyaZ8textImplFNaNbNfAyakAyaZAya':
: undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
Error: linker exited with status 1
```

**Comment by DmitryOlshansky** [comment]
Let's hope it likes it now.

**Comment by wilzbach** [comment]
> Let's hope it likes it now.

```
OS: linux
generated/d_do_test.o: In function `_D3std4conv29__T8textImplTAyaTAyaTkTAyaTkZ8textImplFNaNbNfAyakAyakZAya':
d_do_test.d:(.text._D3std4conv29__T8textImplTAyaTAyaTkTAyaTkZ8textImplFNaNbNfAyakAyakZAya[_D3std4conv29__T8textImplTAyaTAyaTkTAyaTkZ8textImplFNaNbNfAyakAyakZAya]+0x3d): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
d_do_test.d:(.text._D3std4conv29__T8textImplTAyaTAyaTkTAyaTkZ8textImplFNaNbNfAyakAyakZAya[_D3std4conv29__T8textImplTAyaTAyaTkTAyaTkZ8textImplFNaNbNfAyakAyakZAya]+0x79): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o: In function `_D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv':
d_do_test.d:(.text._D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv[_D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv]+0x38): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result5emptyMFNaNbNdNiNfZb'
d_do_test.d:(.text._D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv[_D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv]+0x4a): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result5frontMFNaNbNdNiNfZa'
d_do_test.d:(.text._D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv[_D3std5array17__T8AppenderTAyaZ8Appender94__T3putTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ3putMFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZv]+0x5b): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result8popFrontMFNaNbNiNfZv'
generated/d_do_test.o: In function `_D3std4conv17__T6toImplTAyaTkZ6toImplFNaNbNekkE3std5ascii10LetterCaseZAya':
d_do_test.d:(.text._D3std4conv17__T6toImplTAyaTkZ6toImplFNaNbNekkE3std5ascii10LetterCaseZAya[_D3std4conv17__T6toImplTAyaTkZ6toImplFNaNbNekkE3std5ascii10LetterCaseZAya]+0x51): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o: In function `_D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa':
d_do_test.d:(.text._D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa[_D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa]+0x10): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result6lengthMFNaNbNdNiNfZk'
d_do_test.d:(.text._D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa[_D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa]+0x4b): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result5emptyMFNaNbNdNiNfZb'
d_do_test.d:(.text._D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa[_D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa]+0x57): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result5frontMFNaNbNdNiNfZa'
d_do_test.d:(.text._D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa[_D3std5array96__T5arrayTS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZ5arrayFNaNbNfS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6ResultZAa]+0x8c): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result8popFrontMFNaNbNiNfZv'
generated/d_do_test.o: In function `_D3std4conv36__T8textImplTAyaTAyaTkTAyaTykTAyaTkZ8textImplFNaNbNfAyakAyaykAyakZAya':
d_do_test.d:(.text._D3std4conv36__T8textImplTAyaTAyaTkTAyaTykTAyaTkZ8textImplFNaNbNfAyakAyaykAyakZAya[_D3std4conv36__T8textImplTAyaTAyaTkTAyaTykTAyaTkZ8textImplFNaNbNfAyakAyaykAyakZAya]+0x3d): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
d_do_test.d:(.text._D3std4conv36__T8textImplTAyaTAyaTkTAyaTykTAyaTkZ8textImplFNaNbNfAyakAyaykAyakZAya[_D3std4conv36__T8textImplTAyaTAyaTkTAyaTykTAyaTkZ8textImplFNaNbNfAyakAyaykAyakZAya]+0xb5): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o: In function `_D3std4conv30__T8textImplTAyaTAyaTkTAyaTykZ8textImplFNaNbNfAyakAyaykZAya':
d_do_test.d:(.text._D3std4conv30__T8textImplTAyaTAyaTkTAyaTykZ8textImplFNaNbNfAyakAyaykZAya[_D3std4conv30__T8textImplTAyaTAyaTkTAyaTykZ8textImplFNaNbNfAyakAyaykZAya]+0x3d): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o: In function `_D3std4conv23__T8textImplTAyaTAyaTkZ8textImplFNaNbNfAyakZAya':
d_do_test.d:(.text._D3std4conv23__T8textImplTAyaTAyaTkZ8textImplFNaNbNfAyakZAya[_D3std4conv23__T8textImplTAyaTAyaTkZ8textImplFNaNbNfAyakZAya]+0x3d): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
generated/d_do_test.o: In function `_D3std4conv27__T8textImplTAyaTAyaTkTAyaZ8textImplFNaNbNfAyakAyaZAya':
d_do_test.d:(.text._D3std4conv27__T8textImplTAyaTAyaTkTAyaZ8textImplFNaNbNfAyakAyaZAya[_D3std4conv27__T8textImplTAyaTAyaTkTAyaZ8textImplFNaNbNfAyakAyaZAya]+0x3a): undefined reference to `_D3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZS3std4conv47__T7toCharsVhi10TaVE3std5ascii10LetterCasei1TkZ7toCharsFNaNbNiNfkZ6Result'
collect2: error: ld returned 1 exit status
Error: linker exited with status 1
Makefile:201: recipe for target 'generated/d_do_test' failed
```

**Comment by MetaLang** [comment]
This has been green for awhile now. How should we move forward with this?

File: std/uni.d [hunk 1]
```diff
@@ -6008,6 +6008,533 @@ template SetSearcher(alias table, string kind)
...
+            val = val * 16 + current - 'A' + 10;
+        else
+            throw new Exception("invalid escape sequence");
+    }
+    enforce(val <= 0x10FFFF, "invalid codepoint");
+    str = str[maxDigit..$];
+    return val;
+}
+
+@system unittest //BUG canFind is system
```
Comment (mb64) [inline std/uni.d]: While it's being moved, I think it's worth noting that `canFind` is no longer system (see its [`@safe` unittest](https://github.com/dlang/phobos/blob/master/std/algorithm/searching.d#L2524)), so this can be changed.

**Comment by DmitryOlshansky** [comment]
> This has been green for awhile now. How should we move forward with this?

From my side, absoultely - `pry` won't build without this patch, which is hardly increasing its usability.
Let's just check what's the matter with circle-ci.

File: std/uni.d [hunk 2]
```diff
@@ -622,7 +622,7 @@ import std.range.primitives; // back, ElementEncodingType, ElementType, empty,
     // save
 import std.traits; // isConvertibleToString, isIntegral, isSomeChar,
     // isSomeString, Unqual
-
+import std.exception; // enforce
```
Comment (wilzbach) [inline std/uni.d]: Please use selective imports to avoid symbol leakage.

File: std/uni.d [hunk 3]
```diff
@@ -6156,6 +6707,84 @@ template SetSearcher(alias table, string kind)
...
+        auto usParser = UnicodeSetParser!Range(range, casefold);
+        auto set = usParser.parseSet();
+        range = usParser.range;
+        return set;
+    }
+
+    ///
+    @safe unittest
+    {
+        import std.uni;
```
Comment (wilzbach) [inline std/uni.d]: CI error is:

> Local imports should specify the symbols being imported to avoid hiding local symbols.

File: std/uni.d [hunk 4]
```diff
@@ -6156,6 +6707,84 @@ template SetSearcher(alias table, string kind)
...
+            return front & 0x1f;
+        }
+    }
+
+    //parse and return a CodepointSet for \p{...Property...} and \P{...Property..},
+    //\ - assumed to be processed, p - is current
+    static package CodepointSet parsePropertySpec(Range)(ref Range p,
+        bool negated, bool casefold)
+    {
+        import std.ascii;
```
Comment (wilzbach) [inline std/uni.d]: CI error

> Local imports should specify the symbols being imported to avoid hiding local symbols.

File: std/uni.d (hunk 2 above)
Comment (DmitryOlshansky) [inline std/uni.d]: Last time I checked it was the other way around

Comment (wilzbach) [inline std/uni.d]: Yes, but then we discovered that non-selective imports are leaking:

https://github.com/dlang/phobos/pull/5584#issuecomment-314910297

tl;dr: if you add `std.exception` here, people can import all symbols from `std.exception` from `std.uni`

Comment (MetaLang) [inline std/uni.d]: @wilzbach according to the linked comment even selective imports leak so it wouldn't help to change it.

Comment (DmitryOlshansky) [inline std/uni.d]: Oh for **ck sake... I'll change this one but the other will stay as is for now, I don't want to mix concerns in a single pull request that supposed to be refactoring.

**Comment by MetaLang** [comment]
The only other concern is the test coverage (29% seems pretty poor), but I don't know if that's worth delaying merging.

**Comment by DmitryOlshansky** [comment]
> The only other concern is the test coverage (29% seems pretty poor), but I don't know if that's worth delaying merging.

I'm tired to hell of explaining that our FREAKING AMZAING CI will only do a FILE AT A TIME TEST to show coverage. Regex tests live in a separate file, no way a file at a time will catch proper coverage stats.

TL;DR: Our coverage report is horribly broken, not my problem though.

**Comment by MetaLang** [comment]
@DmitryOlshansky fair enough. I'll leave it to @wilzbach to merge.

**Comment by wilzbach** [comment]
> I'm tired to hell of explaining that our FREAKING AMAZING CI will only do a FILE AT A TIME TEST to show coverage. Regex tests live in a separate file, no way a file at a time will catch proper coverage stats.

Yup.

> TL;DR: Our coverage report is horribly broken, not my problem though

Yes :/

**Comment by quickfur** [comment]
This PR caused a -dip1000 regression: https://issues.dlang.org/show_bug.cgi?id=17961

tl;dr: Now it's impossible to compile any D code with `-unittest -dip1000` if `std.uni` is used anywhere in the code.

**Comment by quickfur** [comment]
This PR also caused a regression in compilation times: https://issues.dlang.org/show_bug.cgi?id=18378

:-(

**Comment by wilzbach** [comment]
See https://github.com/dlang/phobos/pull/5981

**Comment by quickfur** [comment]
I'm talking about *compilation speed*, not the performance of the regex code itself.  Both need to be fixed. :-P
